src/models/generalized_mixture_reach_surface.py

```diff
@@ -0,0 +1,186 @@
+          r: a length n list of length p vectors. Single publisher reaches for
+            each training point.
+
+        Returns:
+          A length n list of length c vectors that indicate each
+          theta_j^{(k)}.
+        """
+        theta = []
+        for k in range(self._n):
+            theta_k = np.zeros(self._c)
+            for j in range(self._c):
+                x = a[[j + i * self._c for i in range(self._p)]] * r[k] / self._N
+                theta_k[j] = (1 - np.prod(1 - x)) / np.sum(x)
+            theta.append(theta_k)
+        return theta
+
+    def solve_a_given_z(self, z, show_status=False):
+        """A key intermediate step to fit GM model.
+
+        Args:
+          z: a length-n list of length-(p * c) vectors.
+
+        Returns:
+          a length-(p * c) vector of the optimization problem above
+
+          𝑃=∑𝑛𝑘=1𝑧(𝑘)[𝑧(𝑘)]⊤ ,  𝑞=−∑𝑛𝑘=1𝑦(𝑘)𝑧(𝑘) ,  𝐺=− np.eye( 𝑝×𝑐 ),
+          ℎ= np.zeros(𝑝×𝑐), 𝐴=np.kron(np.eye(c), np.ones(p)), 𝑏=np.ones(c),
+          𝑎=(𝑎11,…,𝑎1𝑐,…,𝑎𝑝1,…,𝑎𝑝𝑐), 𝑧=(𝑧11,…,𝑧1𝑐,…,𝑧𝑝1,…,𝑧𝑝𝑐).
+        """
+        P = matrix(
+            sum(
+                [
+                    np.matmul(
+                        z[k].reshape(self._p * self._c, 1),
+                        z[k].reshape(1, self._p * self._c),
+                    )
+                    for k in range(self._n)
+                ]
+            )
+        )
+        q = -matrix(sum(self._data[k].reach() * z[k] for k in range(self._n)))
+        G = -matrix(np.eye(self._p * self._c))
+        h = matrix(np.zeros(self._p * self._c))
+        A = matrix(np.kron(np.eye(self._p), np.ones(self._c)))
+        b = matrix(np.ones(self._p))
+        res = solvers.qp(P, q, G, h, A, b)
+        if show_status:
+            print(res["status"])
+        return np.array(res["x"]).flatten()
+
+    def update_a_from_theta(self, theta, r):
+        """Udpate theta.
+
+        Args:
+          theta: a length n list of length c vectors. Current best guess of
+            theta.
+          r: a length n list of length p vectors. Single publisher reaches for
+            each training point.
+
+        Returns:
+          A length (p * c) vector indicating each $a_{ij}$.
+        """
+        z = self.get_z_from_theta(theta, r)
+        return self.solve_a_given_z(z)
+
+    def get_z_from_theta(self, theta, r):
+        z = []
+        for k in range(self._n):
+            z_k = np.zeros(self._p * self._c)
+            for i in range(self._p):
+                for j in range(self._c):
+                    z_k[j + i * self._c] = r[k][i] * theta[k][j] / self._N
+            z.append(z_k)
+        return z
+
+    def _fit(self) -> None:
+        reach_vectors = np.array(
+            [
+                self.get_reach_vector(reach_point.impressions)
+                for reach_point in self._data
+            ]
+        )
+        cur_a = np.random.dirichlet(np.ones(self._c) / 5, size=self._p).flatten()
```

Comment:
If using randomness, there needs to be a way to set the random seed.

Reply:
Good point